Random input selection impl

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#92 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have implemented random input selection 
- [x] I have added respective unit and property tests 


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on test/integration/Test/Integration/APISanityChecks.hs:
```diff
@@ -1,132 +0,0 @@
-{-# OPTIONS_GHC -fno-warn-orphans #-}
```
So, I've entirely removed that one. It was... buggy and I am actually not even sure we use to run it before. 
I like the idea however and we might reconsider later. But for now, let's leave it out.

KtorZ commented on test/integration/Test/Integration/Framework/Cluster.hs:
```diff
@@ -0,0 +1,182 @@
+module Test.Integration.Framework.Cluster
```
So hopefully, a lot of this will go when we achieve full decoupling. Here, we still need to hijack an edge node's identity and sort of re-write the `startCluster` method normally provided by the cluster package.

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -329,13 +301,13 @@ walletName =
 expectFieldEqual getter a = \case
     Left e  -> wantedSuccessButError e
-    Right s -> a `shouldBe` view getter s
+    Right s -> view getter s `shouldBe` a
```
The error outputs renders better when we put what we expect first :)

KtorZ commented on test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -569,16 +520,20 @@ withNextFaucet actionWithFaucet = do
+        \\n\nspec: &default_core_genesis_spec\
+        \\n\tinitializer:\
+        \\n\t\ttestBalance:\
+        \\n\t\t\tpoors: ???\n"
```
Gotta love error messages :)

KtorZ commented on test/integration/Test/Integration/Framework/Scenario.hs:
```diff
@@ -42,19 +41,7 @@ instance Example (Scenario s IO ()) where
+            let acquire = takeMVar mvar
+            let release = putMVar mvar
+            let between = runStateT io >=> (putMVar mvar . snd)
+            bracketOnError acquire release between
```
So, when a test fails, the underlying `io` actually throws. So if we don't put the `MVar` back, we're screwed because the next one holds forever.

KtorZ commented on test/integration/Test/Integration/Framework/Scenario.hs:
```diff
@@ -42,19 +41,7 @@ instance Example (Scenario s IO ()) where
-    mapItem item = item
-        { itemExample =
-            \p action -> (itemExample item) p (action . mapActionWith)
-        }
```
now it's just `beforeAll (newMVar ctx)` o/

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -12,7 +12,7 @@ spec :: Scenarios Context
     scenario "successful payment appears in the history" $ do
         fixture <- setup $ defaultSetup
-            & initialCoins .~ [100]
+            & initialCoins .~ [1000000]
```
Kinda forgot about fees when I wrote this.

KtorZ commented on test/integration/Test/Integration/Scenario/Wallets.hs:
```diff
@@ -42,18 +43,18 @@ spec = do
             ]
 
-    scenario "UTxO statistics reflect wallet's activity" $ do
+    xscenario "UTxO statistics reflect wallet's activity" $ do
```
So this one fails and has a weird output.. Only the tx above `1000` figures in the stats... I couldn't quite figure out why at first glance so I left it out for now and will open a corresponding ticket.

KtorZ commented on test/integration/configuration.yaml:
```diff
@@ -0,0 +1,143 @@
+            richmen:      3
+            richmenShare: 0.99
+            useHDAddresses: True
+            totalBalance: 11111111111111110000
```
So, this is roughly the part we're interested in. This gives us `100` faucet wallets which all have roughly 300B Ada each...  if we need more faucets (one per scenario), we can simply increase `poors`, and if we need more ADA per faucet, we can increase `totalBalance` or reduce `richmenShare`. Note that generating `poors` isn't "free" of time, because we have to generate all the corresponding keys (hence the little time before the test get started!). The more we have, the longer it is. For know set to 100, and when we'll lack of it, it fails with a rather explicit error (cf above).

KtorZ commented on test/integration/configuration.yaml:
```diff
@@ -0,0 +1,143 @@
+          seed: 0
+        blockVersionData: &default_core_genesis_spec_blockVersionData
+          scriptVersion:     0
+          slotDuration:      5000
```
Gotta see down to what we can go here :stuck_out_tongue:  Tests ran in less than 10 minutes on my machine which is rather okay I believe. The delay are usually when we wait after transactions to be validated; by reducing the slot time, we reduce the delay.

KtorZ commented:
Merging as is. Will open another PR with documentation ASAP :+1:

paweljakubas committed:
Random input selection impl

Add NotEnoughMoney and UtxoNotEnoughFragmented errors

paweljakubas commented on src/Cardano/Wallet/CoinSelection/Random.hs:
```diff
@@ -0,0 +1,215 @@
+    pure res
+
+
+-- Selecting coins to cover at least the specified value
```
This is how I understand how this algorithm operates

paweljakubas commented on src/Cardano/Wallet/CoinSelection/Random.hs:
```diff
@@ -0,0 +1,215 @@
+        withIx m' ix = Just (Map.elemAt ix m', Map.deleteAt ix m')
+
+
+instance MonadRandom m => MonadRandom (ExceptT e m) where
```
need this unfortunately

akegalj commented on src/Cardano/Wallet/CoinSelection/Random.hs:
```diff
@@ -0,0 +1,227 @@
+            $ NE.sortBy (flip $ comparing coin) txOutputs
+    let n = maximumNumberOfInputs opt
+
+    let moneyRequested = sum $ (getCoin . coin) <$> txOutputsSorted
```
I think parens can be dropped here 
```haskell
sum $ getCoin . coin <$> txOutputsSorted
```

paweljakubas replied:
:+1:

paweljakubas committed:
Use fallback not per step but optionally in the end

paweljakubas committed:
Add unit tests

paweljakubas committed:
Add property tests

Don't use Identity with MonadRandom

Fix maxNumberOfInput in propertyFragmentation

Hush hlint

paweljakubas commented on test/unit/Cardano/Wallet/CoinSelection/RandomSpec.hs:
```diff
@@ -0,0 +1,290 @@
+                })
+
+    describe "Coin selection properties : Random algorithm" $ do
+        xit "forall (UTxO, NonEmpty TxOut), \
```
it stucks when `random` is called, probably because of wrong impl of MonadRandom instance of Identity at the end of this file

paweljakubas commented on test/unit/Cardano/Wallet/CoinSelection/RandomSpec.hs:
```diff
@@ -0,0 +1,290 @@
+            <*> vectorOf n arbitrary
+        return $ UTxO $ Map.fromList utxo
+
+instance MonadRandom Identity where
```
this is probably wrong

akegalj replied:
it makes sense to me that this will loop forever. On the right side you are calling the same `getRandomBytes` defined in the `instance MonadRandom Identity` , with the same params. 

Looking at https://www.stackage.org/haddock/lts-13.15/cryptonite-0.25/Crypto-Random-Types.html what I believe you should have aimed for is to use one of two implementations of `class DRG` (from the same module) for implementation of `MonadRandom Identity` as you would like to use some deterministic pseudo random generator in this pure code.

I can have a closer look later and experiment if this doesn't give any clue (hope its the right one)

paweljakubas replied:
thanks @akegalj for the comment. Indeed, random monad within random algo needs more effects power than the monad within largestFirst. there is a need of more effects to enable randomness - be in IO or have initial seed/generator.

KtorZ committed:
remove extra space in module export

KtorZ committed:
remove unecessary orphan instance

KtorZ committed:
update docstring comment for 'random' selection to match the specification

KtorZ committed:
add comments to 'TargetRange' type

KtorZ committed:
replace 'Coin' with 'Word64' in the 'TargetRange' type

We don't need these various targets to be 'Coin', some may actually not
be; in the end, we will select several coins, so it's not impossible
that the sum of their value will overflow the max coin value. In the
rare case this would happen, we can simply fallback to splitting the
output into two change coins.

KtorZ committed:
use of 'MaybeT' to simplify writing of 'processTxOut'

KtorZ committed:
replace 'Map TxIn TxOut' with 'UTxO' to ease transaction signatures

KtorZ committed:
re-indent where clause on a 4-space indent (instead of 8-space)

KtorZ committed:
remove currently useless fee-related options and types definitions

KtorZ committed:
some minor naming revision: in particular 'atLeast' -> 'coverRandomly'

I found calling the latter 'atLeast' quite confusing with the 'largestFirst' algorithm.

paweljakubas commented on src/Cardano/Wallet/CoinSelection/Random.hs:
```diff
@@ -53,24 +53,45 @@ data TargetRange = TargetRange
+--    step.)
+--
+-- 2. Randomly select outputs from the UTxO, considering for each output if that
+--    output is animprovement. If it is, add it to the transaction, and keep
```
**animprovement** -> an improvement

paweljakubas commented on src/Cardano/Wallet/CoinSelection/Random.hs:
```diff
@@ -202,29 +200,20 @@ processTxOut maxNumInputs input txout =
-            targetAim <-
-                if isValidCoin (Coin $ 2*v) then Just (Coin $ 2*v) else Nothing
-            targetMax <-
-                if isValidCoin (Coin $ 3*v) then Just (Coin $ 3*v) else Nothing
```
we are not afraid of situation that for instance targetMax will exceed upperBound of what Coin can have???

KtorZ committed:
factor out common test logic in largest first and random coin selection

KtorZ committed:
get system DRG through Hspec bracket mechanism (instead of unsafe IO)
